src/framebuffer.ml

```diff
@@ -12,15 +15,15 @@ let init (dimensions : int * int) (f : int -> int -> int) : t =
   let width, height = dimensions in
   if width <= 0 then raise (Invalid_argument "Invalid width");
   if height <= 0 then raise (Invalid_argument "Invalid height");
-  Array.init height (fun y ->
-    Array.init width (fun x ->
-        f x y
-      )
-  )
-
-let pixel_write (x : int) (y : int) (col : int) (buffer : t) =
-  if (x >= 0) && (x < Array.length (buffer.(0))) && (y >= 0) && (y < Array.length buffer) then
-    buffer.(y).(x) <- col
+  { buffer = Array.init height (fun y ->
+      Array.init width (fun x -> f x y));
+    dirty = false }
+
+let pixel_write (x : int) (y : int) (col : int) (fb : t) =
+  if (x >= 0) && (x < Array.length (fb.buffer.(0))) && (y >= 0) && (y < Array.length fb.buffer) then begin
+    fb.buffer.(y).(x) <- col;
+    fb.dirty <- true  (* Mark framebuffer as dirty *)
```

Comment:
This isn't the only place we modify a Framebuffer.t - we need to ensure that all paths that lead to the buffer being modified cause the dirty flag to be set true. Some of the methods do call pixel_write, and those you've covered, but a bunch more do not for performance reasons, and those you'll need to update.

The way to then get confidence we've caught them all is to update the unit tests in `test_primitives.ml` to check that the dirty bit is always  not set before drawing, and then set after drawing (and not unset by drawing again :)